docs/design/federated-services.md

```diff
@@ -0,0 +1,514 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+## Cross-cluster Load Balancing and Service Discovery 
+### Requirements and System Design
+### by Quinton Hoole, Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in
+   a Kubernetes cluster) must be able to easily discover and connect
+   to endpoints for Kubernetes services on which they depend in a
+   consistent way, irrespective of whether those services exist in a
+   different kubernetes cluster within the same cluster federation.
+   Hence-forth referred to as "cluster-internal clients", or simply
+   "internal clients".
+1. **External discovery and connection**: External clients (running
+   outside a Kubernetes cluster) must be able to discover and connect
+   to endpoints for Kubernetes services on which they depend.
+   1. **External clients predominantly speak HTTP(S)**: External
```

Comment:
What about the use case for clients that are external to the cluster but still inside to the same internal network? e.g. VMs on GCE/AWS or other machines on prem for a bare metal install. Those clients may not need to use https and we want a way to allow them to connect to services without needing to create an endpoint that is visible to the internet.